Stateless observation hydra node

v0d1ch opened the pull request:
<!-- Describe your change here -->


TODO:
- [x] Make E2E tests green with multiple heads on a network
- [x] `fromPlutusCurrencySymbol` partial
- [x] FIXME in broken `Greetings`
- [x] FIXME on comparing observations in HandlersSpec.hs
- [x] Test `observe` functions with multiple Head UTxO
- [ ] Actually get rid of `Hydra.Chain.Direct.State` module from production code (use it only to generate transactions and rename it)

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [ ] Haddocks updated or not needed
* [ ] No new TODOs introduced or explained herafter

v0d1ch committed:
Use observeHeadTx instead of observeSomeTx in Chain.Direct.Handlers

ch1bo committed:
Implement a convertObservation to match OnHeadTx interface

ch1bo committed:
Switch ChainStateAt to just hold a UTxO

This is mocking out the usage sites of the chain state on the
transaction creation site.

ffakenz committed:
Mock out ChainState usage in prepareTxToPost

v0d1ch committed:
Change the interface of the abort function

We started to work on a "spendable" utxo

v0d1ch committed:
Create HeadSeed type and add it to AbortTx

ffakenz committed:
Added bytestring seed to HeadSeed type along with serialization instances

ch1bo committed:
Define a property test for HeadSeed <-> TxIn

This also mocks usages of 'abort' in hydra-node tests.

v0d1ch committed:
Implement roundtrip from HeadSeed -> TxIn

ffakenz committed:
Predict HeadSeed on e2e spec

This is not a convenient solution.
Ultimately we would get the seed from an OnInitTx obeservation.

v0d1ch committed:
Wip start with finding correct outputs

ffakenz committed:
Fix compilation of tx-cost with new abort signature

ffakenz committed:
Add new post tx error to fail when an abort tx cannot be built

We expect bob to be able to construct the abort tx but failing to submit it.

ch1bo committed:
Only try to construct abort tx after observing init

v0d1ch committed:
Make invocations of abort compile in test-suites

This follows the pattern of unsafeCommit and likely will be changed more
in the future (e.g. move all unsafe and generator code into tests)

ch1bo committed:
Move HeadSeed and rename constructors

Also adds testHeadId and testHeadSeed as fixtures to tests.

ch1bo committed:
Draft DirectChainSpec to take the seed from OnInitTx

ffakenz committed:
Add HeadSeed to onInitTx

So to be able to observe the seed from any init tx.

ffakenz committed:
Use correct seed in abort state spec

v0d1ch committed:
Use correct seed in StateSpec to make the test pass

v0d1ch committed:
Drop the unneeded test in the StateSpec

- Any ignoring of the observations should now be in the higher levels of
  protocol.

ch1bo committed:
Create a behavior test to assert nodes ignore init txs they are not part of

This was handled before in the chain layer, but now it is expected to be
in the protocol logic.

We also update signature of IgnoredHeadInitializing to report on the
off-chain identities of a seen Head initTx. This should be equivalent
for most users, but allows us to reason about it in the behavior
logic (where there are no on-chain identities / keys)

ffakenz committed:
Moved checking of the init tx to the HeadLogic

ffakenz committed:
Keep track of the head seed in the head state

ch1bo committed:
Add abort property test to not mix inputs

This test ensures that abort tx creation is correctly selecting inputs
if the known, spendable UTxO contains more than only "this heads" UTxO.

ch1bo committed:
Fix abort tx creation to select the right UTxO

This is now using only UTxO holding head tokens of the head to abort.

ch1bo committed:
Remove unused findTxOutsByScript

ch1bo committed:
Remove the need for NetworkId from observeRawInitTx

This is not needed if we do not compare the full address, but only the
payment credential via `isScriptTxOut`

ch1bo committed:
Migrate tests about ingored heads to HeadLogicSpec

It's the protocol logic's responsibility now to ignore heads, so we can
test it in the HeadLogicSpec already.

ch1bo committed:
Fix head ignoring in HeadLogic

What we want to do to observe proper init tx is:
 - Compare parties and make sure they are the same
 - Compare contestation period and make sure it matches
 - Make sure we are part of configured peers

v0d1ch committed:
Fix api schema for IgnoredHeadInitializing

v0d1ch committed:
Make the NodeSpec tests pass

I had to use the default value for the contestation period
which is 60s. We should revisit and see if we can remove cperiod
from fixtures and just use defaultContestationPeriod everywhere.

v0d1ch committed:
Replace observeSomeTx with observeHeadTx in StateSpec

v0d1ch committed:
Remove observeSomeTx since it is replaced by observeHeadTx

Use observeHeadTx everywhere in tests and derive missing Eq and Show instances

v0d1ch committed:
Fix the golden files

Since we renamed participants to parties these need to be altered too

v0d1ch committed:
Change commit tx creation to use UTxO

ch1bo committed:
Moved PolicyId functions into respective model in hydra-cardano-api

ch1bo committed:
Add a headIdToPolicyId conversion and use it in commit

ch1bo committed:
Remove undefined from commit tx construction invocation

ch1bo committed:
Use headStatus read model to implement /commit

We can use the local "projection" of headStatus as a source for deciding
whether a commit it possible and to get hold of the head id.

This is currently breaking the Greetings message API, because the same
data type is used to fill the 'headStatus' there.

ffakenz committed:
Use the new HeadStatus scheme to fix the deserialization of Greetings message

ffakenz committed:
Fix close tx creation

ffakenz committed:
Keep track of the head seed in the head open state

ffakenz committed:
Add collected utxo to open a head from collect-com tx observation

ffakenz committed:
Minor fix on init observation

Add missing head seed attribute from seed tx input.

v0d1ch committed:
Find the head utxo to spend in the close tx

Add error type to return from a close function since it now returns an
Either
Add and use unsafeClose function in the tests

ffakenz committed:
Resolving utxo hash when closing a head on initial state

By using the hash of the utxo inside the initial snapshot.

Also removed some redundant imports.

v0d1ch committed:
Make DirectChainSpec compile

ffakenz committed:
Remove utxo from CloseTx

We can use the UTxO from the chain state like we do it in the other transactions.

ffakenz committed:
Remove head seed from CloseTx

We can use the head id instead.

ffakenz committed:
Add headId to contest post chain tx

Also use the chain state utxo instead of the closed direct chain state.

v0d1ch committed:
CollectCom: Add headId parameter to collect

ffakenz committed:
Throw CannotFindHeadOutputToContest contest error while preparing ContestTx

Also fixed a warning about missing headId to be passed to ContestTx on HeadLogic.

v0d1ch committed:
CollectCom: add UTxO parameter

ffakenz committed:
Fix genStClosed by taking the head id from the open state instead of using arbitrary

v0d1ch committed:
CollectCom propagate head utxo

Now the head input and output are comming in from our `spendableUTxO`
and we removed it from InitialThreadOutput.

collect now returns an Either so this change is propagated throughout.

v0d1ch committed:
CollectCom: Find commit outputs in the 'spendableUTxO'

v0d1ch committed:
CollectCom: Use contestation period from 'ChainContext'

v0d1ch committed:
CollectCom: Get parties from ChainContext

Now InitialState is no longer needed

v0d1ch committed:
Fix a small typo in the test description

v0d1ch committed:
CollectCom: Pass head address instead of network id

v0d1ch committed:
CollectCom: Improve haddocks

Also rename allParties to allParties' when it comes to local name
for hydra context parties.

v0d1ch committed:
Contest: check and extract information from the head datum

ffakenz committed:
Remove utxo from CollectComTx

v0d1ch committed:
Remove HeadParameters from CloseTx

ffakenz committed:
Fanout: propagate head utxo & check + extract information from the head datum

ffakenz committed:
Fix cluster specs after last changes on PostChainTx constructors

ffakenz committed:
Fix missing arg warn in HeadLogicSpec

Also applied some minor formatting changes.

ffakenz committed:
Fanout: simplify head datum check

v0d1ch committed:
Extract helper to get the head utxo

v0d1ch committed:
Fix DirectChainSpec test related to observing close tx

ffakenz committed:
Removed unused type OnChainId

ffakenz committed:
Rename ChainStateAt utxo field to spendableUTxO

ffakenz committed:
Resolve TODO as it is not convenient to remove the parameters from the OpenState

Due to it is being used by the Node to validate any mismatch configuration.

ffakenz committed:
Replace fixture cperiod by the default option

v0d1ch committed:
Fix the generator for ContestationPeriod in CollectCom mutation

v0d1ch committed:
Use correct utxo for unsafeClose and unsafeCollect

v0d1ch committed:
Use correct utxo in StateSpec tests

v0d1ch committed:
Fix golden files

v0d1ch committed:
Add own party to contested parties on contest

Also use arbitrary utxo for contest snapshot

ffakenz committed:
Update logs.yaml and fix hydra logging spec

ffakenz committed:
More fixes to api and logs schema

v0d1ch committed:
Do not alter the closedParties in contest function

v0d1ch committed:
)Add helper function to look at the head datum

v0d1ch committed:
Fix collect tx creation

We should not re-arrange parameters. In fact, all the head specific
information in the ChainContext should be removed eventually as the
chain layer will become head independent.

ch1bo committed:
Refactor collectCom to take HeadParameters

This way we have less error cases in the tx construction.

ffakenz committed:
Pass missing arguments fix

ffakenz committed:
Pass head params as part of CostTx

to remove unnecesary error types when checking over the given UTxO.

ffakenz committed:
Fix close validates within maxTxExecutionUnits

We were generating a snapshot without being signed by all parties

ffakenz committed:
Remove cp from chain context

ffakenz committed:
Remove other parties from chain context

ffakenz committed:
Remove other parties from chain context

ffakenz committed:
Remove peer verification keys from chain context

v0d1ch committed:
Read verification keys from file properly

v0d1ch committed:
Include our vkey also

v0d1ch committed:
Generate commit for specific actor

v0d1ch committed:
Add script registry utxo for unsafeCommit and genCommitTxMutation

v0d1ch committed:
Fix golden files

v0d1ch committed:
Lookup redeemer against the resolved utxo

v0d1ch committed:
Add missing fields for the shrinker after rebase

ch1bo commented on hydra-node/src/Hydra/Chain/Direct.hs:
```diff
@@ -183,12 +177,19 @@ withDirectChain tracer config ctx wallet chainStateHistory callback action = do
 
+  -- read other parties cardano vkeys
+  otherCardanoKeys <- mapM (readFileTextEnvelopeThrow (AsVerificationKey AsPaymentKey)) cardanoVerificationKeys
+  -- obtain our vkey from reading our skey from file
```
Could: remove _obvious_ commentary

v0d1ch committed:
Fix the headIsInitializingWith to correctly check the set of parties

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -1115,3 +1143,13 @@ findHeadAssetId txOut =
+  datum <- txOutScriptData $ toTxContext headOutput
+  fromScriptData datum
+ where
+  headScript = fromPlutusScript Head.validatorScript
```
Must: remove unused code

ffakenz replied:
i think the comment got outdated.

it is being used.

ch1bo replied:
This function is not used

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -160,20 +173,18 @@ onInitialChainCommitTx st newChainState pt utxo =
   postCollectCom =
     OnChainEffect
-      { postChainTx = CollectComTx $ fold newCommitted
+      { postChainTx = CollectComTx{headId, headParameters = parameters}
```
Odd that we had constructed the tx with expected commits before, but now we seemingly infer them from the spent UTxO. But on the other hand parameters were added here instead of inferring them from the spendable UTxO.

Also, the `AbortTx` does have the committed UTxO.. this is very inconsistent.

Must: be consistent in one or the other approach

ch1bo commented on hydra-cardano-api/src/Hydra/Cardano/Api/PolicyId.hs:
```diff
@@ -2,7 +2,24 @@ module Hydra.Cardano.Api.PolicyId where
+-- | Convert a plutus 'CurrencySymbol' into a cardano-api 'PolicyId'.
+fromPlutusCurrencySymbol :: CurrencySymbol -> PolicyId
+fromPlutusCurrencySymbol = PolicyId . unsafeScriptHashFromBytes . fromBuiltin . unCurrencySymbol
+{-# DEPRECATED fromPlutusCurrencySymbol "this is partial" #-}
```
Should: address this and make it total

ffakenz replied:
I attempted to address this issue, but I found myself uncertain about how to handle this function when calling it from `headIdToPolicyId`. This function is utilized to prepare all our transactions in module `Hydra.Chain.Direct.State`.

As far as I understand, we need to include an error handler as part of `headIdToPolicyId` and then handle it appropriately within the callers. This way, if this function throws an error, we can convert it to the correct TxError.

What do you think?
Should we tackle this as part of this PR, or should we leave it for later?

ch1bo commented on hydra-cluster/test/Test/DirectChainSpec.hs:
```diff
@@ -235,21 +246,21 @@ spec = around (showLogsOnFailure "DirectChainSpec") $ do
             someUTxO <- seedFromFaucet node aliceExternalVk 1_000_000 (contramap FromFaucet tracer)
+            let params = HeadParameters cperiod [alice]
+            postTx $ InitTx params
+            (headId, headSeed) <- aliceChain `observesInTimeSatisfying` hasInitTxWith cperiod [alice]
```
Could: use `HeadParameters` also in `hasInitTxWith`

ch1bo commented on hydra-node/src/Hydra/Chain.hs:
```diff
@@ -146,6 +142,11 @@ data PostTxError tx
+  | FailedToConstructCloseTx
+  | FailedToConstructContestTx
+  | FailedToConstructCollectTx
+  | FailedToConstructFanoutTx
```
Should: Include the individual sub-types to give information about why it failed to construct

ch1bo commented on hydra-node/src/Hydra/Chain.hs:
```diff
@@ -84,19 +79,19 @@ instance IsTx tx => Arbitrary (PostChainTx tx) where
 -- possible to simplify observing the chain.
 data OnChainTx tx
-  = OnInitTx {headId :: HeadId, contestationPeriod :: ContestationPeriod, parties :: [Party]}
+  = OnInitTx {headId :: HeadId, headSeed :: HeadSeed, contestationPeriod :: ContestationPeriod, parties :: [Party]}
```
Could: Make `OnInitTx` a bit more consistent with its `AbortTx` and `CollectComTx` "continuations" by making it take a `headParameters :: HeadParameters` as well.

ffakenz replied:
not sure if i like the result out of it, but done!

ch1bo commented on hydra-node/test/Hydra/HeadLogicSpec.hs:
```diff
@@ -424,7 +428,8 @@ spec =
             deadline = arbitrary `generateWith` 42
             closeTxEvent = observationEvent $ OnCloseTx testHeadId 0 deadline
-            contestTxEffect = chainEffect $ ContestTx latestConfirmedSnapshot
+            params = fromMaybe (Prelude.error "No HeadParameters in Idle state") (getHeadParameters s0)
```
Should: avoid `error` by just constructing some `HeadParameters` from fixtures here (the values do not matter as there is no on-chain verification)

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/TxSpec.hs:
```diff
@@ -45,6 +45,12 @@ import Test.QuickCheck.Property (checkCoverage)
+      prop "headSeedToTxIn . txInToHeadSeed === id" $ \txIn -> do
+        let headSeed = txInToHeadSeed txIn
+        headSeedToTxIn headSeed === Just txIn
+          & counterexample (show headSeed)
```
Should: also add an analogous roundtrip test for `HeadId` here

ch1bo commented on hydra-node/src/Hydra/API/ServerOutput.hs:
```diff
@@ -232,7 +233,7 @@ prepareServerOutput ServerOutputConfig{txOutputFormat, utxoInSnapshot} response
 data HeadStatus
   = Idle
-  | Initializing
+  | Initializing {headId :: HeadId} -- FIXME: This breaks the api of 'Greetings', document and fix api.yaml
```
Must: resolve this FIXME

We have three options:
* Accept the breaking change and persist it in the api.yaml
* Not re-use this `HeadStatus` projection, but a create a new read-model which only gives us the `HeadId`
* Break the `/commit` api to require a `HeadId`

v0d1ch replied:
We already broke the api with changes in the `PostChainTx` didn't we?

v0d1ch committed:
Fix tx-cost benchmarks

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Handlers.hs:
```diff
@@ -295,75 +304,124 @@ chainSyncHandler tracer callback getTimeHandle ctx localChainState =
+        >=> ( \case
+                Nothing -> pure ()
+                Just event -> callback event
+            )
```
IMO this is less clear to read as before (and longer due to the parenthesis)

ffakenz replied:
removing parenthesis made it same size and even reduced some characters.

why is it less readable for you?

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Handlers.hs:
```diff
@@ -189,12 +197,12 @@ finalizeTx ::
   ChainContext ->
-  ChainStateType Tx ->
+  UTxO.UTxO ->
   UTxO.UTxO ->
```
Should: document which UTxO is which in haddock (it's an impossible to understand signature otherwise)

ch1bo replied:
Do we need to separate it even?

ffakenz replied:
i think we want to separate them, so to favor documentation... but you are correct; we can combine them all and have the function to take a single argument.

wdyt?

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/StateSpec.hs:
```diff
@@ -148,16 +153,17 @@ spec = parallel $ do
                 property False
-                  & counterexample ("observeSomeTx returned an error: " <> show err)
+                  & counterexample ("observeHeadTx failed to observe transaction: " <> show tx)
+              _ -> property True
```
Should: extend / add another property test here that we can observe transactions in presence of multiple other heads (similar test on line 307)

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/State.hs:
```diff
@@ -194,20 +211,18 @@ instance HasKnownUTxO ChainState where
-  , ownParty :: Party
-  , otherParties :: [Party]
+  , -- FIXME: peerVerificationKeys should not be here
+    ownParty :: Party
```
Must: remove outdated FIXME comment

ffakenz replied:
that comment has a typo.

it should apply to `ownParty`.

ffakenz replied:
I tried this already but I dislike the changes, so I would also prefer to address this in a separate PR.

wdyt?

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/State.hs:
```diff
@@ -323,68 +328,68 @@ instance HasKnownUTxO ClosedState where
+-- | Construct a commit transaction based on known, spendable UTxO and some
+-- arbitrary UTxOs to commit. This does look for "our initial output" to spend
+-- and check the given 'UTxO' to be compatible. Hence, this function does fail
+-- if already committed or if the head is not initializing.
```
Must: update haddocks of all other functions in the same way

Alternative: remove the need for this module altogether and stick with one set of functions to create/observe transactions (the ones in `Hydra.Chain.Direct.Tx` with possible rename to `Hydra.Chain.Tx`)

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/State.hs:
```diff
@@ -473,20 +495,33 @@ collect ctx st = do
-    -- XXX: Not needing anything of the 'InitialSnapshot' is another hint that
-    -- we should not keep track of an actual initial 'Snapshot'
-    InitialSnapshot{} -> CloseWithInitialSnapshot{openUtxoHash}
+    -- REVIEW: Check if this is as good as using the deserialized output from the datum.
```
Yes. This is following the pattern we use here, the logic layer provides off-chain information that is translated into expected transactions. The on-chain verification will deem it good enough if the off-chain information was correct.

ch1bo reviewed (commented):
There are still some things to clean up. All in all, the change looks better than expected. I have left several suggestions, but plan to address them also myself.

There is also quite some refactoring potential and i will try to remove the `Hydra.Chain.Direct.State` module completely from the "production code". That is, only use it to generate transactions in `tx-cost` and the `hydra-node` test suite.

v0d1ch committed:
PostChainTx changes to the api.yaml

ffakenz committed:
Remove reviewed and unnecessary comments

ffakenz committed:
Fix formatting

ch1bo committed:
Match generated observations in HandlersSpec

ch1bo committed:
Introduce a maybeToEither operator in hydra-prelude

ch1bo committed:
Make fromPlutusCurrencySymbol total

This also adds several error cases (which would have had impure
exceptions before) to call sites in the hydra-node.

ffakenz committed:
Extend other heads observation property tests

Add resolve inputs helper function.

v0d1ch committed:
Use new function resolveInputsUTxO

v0d1ch committed:
Use resolveInputsUTxO in observing transactions

ch1bo committed:
Check actual transitions in observeTx property

This is not yet checking whether we observe the "right" head by id though.

v0d1ch committed:
Check observed headId's in the observeTx property

ffakenz committed:
Add headparams as part of OnInitTx

ffakenz committed:
Avoid errors in HeadLogicSpec by passing default HeadParameters

ffakenz committed:
use HeadParameters in hasInitTxWith

ffakenz committed:
Minor formatting changes

v0d1ch committed:
Fix the Greetings api schema

v0d1ch committed:
Fix IgnoredHeadInitializing api schema

v0d1ch committed:
Make sure we abort the correct head

To prevent aborting the wrong Head `headId` was added to the
`OnAbortTx`. Logic layer now aborts the head that is seen as
initialized.

ch1bo committed:
Fix logs.yaml schema

The flaky test picked up this error

ch1bo committed:
Remove unused peekIntoHeadDatum

ch1bo committed:
Add HeadLogic tests about ignoring observations of other heads

As we removed those guards from the chain layer, it is important that
the logic layer is doing this filtering.

ffakenz committed:
Add headId to OnContestTx and validate against ClosedState during HeadLogic update

ffakenz committed:
Add headId to OnFanoutTx and validate against ClosedState during HeadLogic update

ffakenz committed:
Fix headlogic for collect com

We were missing to check the headId from the observation with ours within the local state.

ch1bo committed:
Re-instantiate an OnChainId

This is needed to ignore OnInitTx because it was not using the right
on-chain keys.

ch1bo committed:
Remove "non-raw" observeInitTx

We were only using observeRawInitTx and all checking of whether "its the
right head" needs to be in the logic layer now.

This commit does not yet migrate the test for this (which was in TxSpec)

v0d1ch committed:
Revert "Fix the Greetings api schema"

ch1bo committed:
Check OnChainId of participants in logic layer

This was done by the chain layer before, but needs to be done now by the
logic layer. This is the only usage of on-chain identities in the
off-chain Head protocol, but it still is required to retain semantics
from before.

This made initializing the Environment a bit hacky, as it is very
specific to the Cardano configuration.

v0d1ch committed:
Revert 'Use the new HeadStatus scheme to fix the deserialization of Greetings message'

v0d1ch committed:
Introduce HeadId Projection and use it

In order to avoid breaking changes in the Greetings output messages
introduce a new projection. This projection only gives you a HeadId
if the Head is in Initializing state.

ch1bo committed:
Add [OnChainId] to OnInitTx and InitTx

Originally only needed for the observation side (OnInitTx), the
BehaviorSpec actually nudges us into also specifying the 'participants'
on the creation side (InitTx). This feels right, despite it requiring
quite a lot of changes in our test suites, because we can remove the
other cardano keys from the direct chain handle! At some point this will
enable us to open a head on-demand with any other participants.

ch1bo committed:
Re-add participants to IgnoredHeadInitializing

This avoids a breaking change on the API, but keeps additional
information about parties and contestationPeriod. This is helpful as all
three things need to match.

ffakenz committed:
Add utxo to CollectComTx

ch1bo committed:
Fix ModelSpec by using concrete OnChainId

We cannot use "derived" on chain identifiers in the Model tests, because
we are actually validating the transactions and need to sign them
correctly. Hence, the participation tokens need to be match the
verification keys used.

v0d1ch committed:
Start with fixing DirectChainSpec

ch1bo committed:
Add Carol as participant in a DirectChainSpec test

ch1bo committed:
Match participants and parties even when not in same order

v0d1ch committed:
Fix and rename projectInitializingHeadId

We must reset the headId when the head is open or aborted, but keep it
on all other server outputs.

ch1bo committed:
Fix CollectCom logic

This was correct originally, but got re-added incorrectly a few commits back

ch1bo committed:
Resolve warnings

ch1bo committed:
Formatting and minor fixes

Avoid some diff to master and fix comments

ch1bo reviewed (approved):
I am happy with the changes now!

Ideally we would update the CHANGELOG and bump versions of packages accordingly.

Let's merge this!

ch1bo committed:
Still write cardano verification keys to disk in chainConfigFor

While the DirectChainSpec tests would not need it, the EndToEnd
scenarios have a fulll hydra-node running where the chain context and
wallet still do load these files.

v0d1ch committed:
Add OnChainId to the api schema

v0d1ch committed:
Bump affected package versions

v0d1ch committed:
Add changelog entry

v0d1ch committed:
Add headId to OnChainTx

This was the last step to make all e2e tests green (two heads on the
same network turned out to be flaky).

v0d1ch committed:
Fixes after rebasing
